becquerel/core/spectrum.py

```diff
@@ -726,6 +787,72 @@ def downsample(self, f, handle_livetime=None):
                         bin_edges_kev=self.bin_edges_kev,
                         livetime=new_livetime)
 
+    def has_uniform_bins(self):
+        """Test whether the Spectrum has uniform binning.
+
+        This is possibly the best way to test for uniform binning within float
+        precision, after https://github.com/0cjs/py-allthesame. It may also be
+        useful to make this a member, not a method, that gets set at Spectrum
+        initialization.
+
+        TODO: this calls self.bin_widths, which is undefined if not calibrated,
+        but this should work regardless.
+
+        Returns:
+          True/False if all binwidths are equal to within np.finfo(float).eps
+        """
+        iterator = iter(self.bin_widths)
+        x0 = next(iterator, None)
+        for x in iterator:
+            if abs(x/x0 - 1.0) > EPS:
+                return False
+        return True
+
+    def find_bin(self, x):
+        """Find the Spectrum bin that contains a value x.
+
+        If the Spectrum has uniform binning, then we just solve the linear
+        equation between bins and x-axis values. Otherwise, we use bisect_left
+        to solve for the insertion point where x would fit in a list of bin
+        edges, then subtract 1 to get the index of the low edge. For an x
+        equal to the bin low edge, bisect_left returns 0, so take the max to
+        avoid a -1.
+
+        TODO: ensure this works with uncal channels, too
+
+        Args:
+          x: value whose bin to find
+
+        Raises:
+          AssertionError: if x is outside the bin edges or equal to up edge
+
+        Returns:
+          The integer bin number that contains x
+        """
+        assert x >= self.bin_edges_kev[0]
+        assert x < self.bin_edges_kev[-1]
+
+        if self.has_uniform_bins():
+            return int((x - self.bin_edges_kev[0]) / self.bin_widths[0])
+        else:
+            return max(bisect_left(self.bin_edges_kev, x) - 1, 0)
```

Comment:
why not use `numpy.searchsorted(self.bin_edges_kev, x, side='left')`, wouldn't add another dependency, and is probably the fastest in any case. I can't imagine a case where `bin_edges_kev` wouldn't be sorted, and if they are not, we probably should enforce it on creation.

Reply:
Another comment. Also we probably should figure out an answer to all the FIXME's. Didn't look to much into those.

Reply:
Thanks, I didn't know about the numpy solution. I'll also update it to use the `'right'` option, as that makes more sense for an inclusive low edge but exclusive high edge.

Reply:
Go for it, I had the `left` because it was the default, haven't thought much about what makes more sense.